envi/tests/test_arch_arm.py

```diff
@@ -271,49 +814,83 @@ def test_envi_arm_operands(self):
         vw = vivisect.VivWorkspace()
         vw.setMeta("Architecture", "arm")
         vw.addMemoryMap(0, 7, 'firmware', '\xff' * 16384*1024)
         vw.addMemoryMap(0x400000, 7, 'firmware', '\xff' * 16384*1024)
         emu = vw.getEmulator()
-        emu.logread = emu.logwrite = True
-
+        #took out next line while testing routine
+        #mu.logread = 
+        emu.logwrite = True
         badcount = 0
         goodcount = 0
-
-        #emu = archmod.getEmulator()
-        #emu.addMemoryMap(0, 7, 'firmware', '\xff' * 16384*1024)
-        #emu.addMemoryMap(0x400000, 7, 'firmware', '\xff' * 16384*1024)
-
-        for bytez, va, reprOp, iflags, emutests in instrs:
-            op = vw.arch.archParseOpcode(bytez.decode('hex'), 0, va)
-            redoprepr = repr(op).replace(' ','').lower()
-            redgoodop = reprOp.replace(' ','').lower()
-            if redoprepr != redgoodop:
-                raise Exception("FAILED to decode instr:  %.8x %s - should be: %s  - is: %s" % \
-                         ( va, bytez, reprOp, repr(op) ) )
-                badcount += 1
-                raise Exception("FAILED to decode instr:  %.8x %s - should be: %s  - is: %s" % \
-                         ( va, bytez, reprOp, repr(op) ) )
-            self.assertEqual((bytez, redoprepr, op.iflags), (bytez, redgoodop, iflags))
-
-            # test some things
-            if not len(emutests):
-                # if we don't have tests, let's just run it in the emulator anyway and see if things break
-                if not self.validateEmulation(emu, op, (), ()):
-                    goodcount += 1
-                else:
-                    raise Exception( "FAILED emulation:  %s" % op )
-                    badcount += 1
-
-            else:
-                for tdata in emutests:  # dict with 'setup' and 'tests' as keys
-                    setup = tdata.get('setup', ())
-                    tests = tdata.get('tests', ())
-                    if not self.validateEmulation(emu, op, setup, tests):
-                        goodcount += 1
+        for archz, bytez, va, reprOp, iflags, emutests in instrs:
+            ranAlready = False  # support for run once only
+            #itterate through architectures - "mask" value is a power of 2 so set that up too
+            for arch_mask in range(number_of_archs):
+                test_arch = int(pow(2,arch_mask)) #mask for architecture to test
+                if ((not ranAlready) or (not self.armTestOnce)) and ((archz & test_arch & self.armTestVersion) != 0):
+                    ranAlready = True
+                    arm.ThumbModule.archVersion = arm.ArmModule.archVersion = archBitMask[arch_mask][0]
+                    op = vw.arch.archParseOpcode(bytez.decode('hex'), 0, va)
+                    redoprepr = repr(op).replace(' ','').lower()
+                    redgoodop = reprOp.replace(' ','')
+                    if redoprepr != redgoodop:
+                        print  bytez,redgoodop
```

Comment:
Understood.  And expected. I have been incrementally cleaning this up as I go along.  Trying to only keep what makes sense to help with the creation and debugging of new code.

Date: Fri, 3 Jun 2016 06:01:24 -0700
From: notifications@github.com
To: vivisect@noreply.github.com
CC: redheadedrod@hotmail.com; author@noreply.github.com
Subject: Re: [atlas0fd00m/vivisect] Update unit tests for opcodes (#3)

In envi/tests/test_arch_arm.py:

> -                    tests = tdata.get('tests', ())
> -                    if not self.validateEmulation(emu, op, setup, tests):
> -                        goodcount += 1
> +        for archz, bytez, va, reprOp, iflags, emutests in instrs:
> +            ranAlready = False  # support for run once only
> +            #itterate through architectures - "mask" value is a power of 2 so set that up too
> +            for arch_mask in range(number_of_archs):
> +                test_arch = int(pow(2,arch_mask)) #mask for architecture to test
> +                if ((not ranAlready) or (not self.armTestOnce)) and ((archz & test_arch & self.armTestVersion) != 0):
> +                    ranAlready = True
> +                    arm.ThumbModule.archVersion = arm.ArmModule.archVersion = archBitMask[arch_mask][0]
> +                    op = vw.arch.archParseOpcode(bytez.decode('hex'), 0, va)
> +                    redoprepr = repr(op).replace(' ','').lower()
> +                    redgoodop = reprOp.replace(' ','')
> +                    if redoprepr != redgoodop:
> +                        print  bytez,redgoodop

before we're done, this will all want to be seriously cleaned up (comments, prints, etc...)  but for now, this suits our purposes.  i only state this because the PR makes it more official.


—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.